topgun/k8s/worker_lifecycle_test.go

```diff
@@ -8,109 +8,129 @@ import (
+		)
 
-			waitAllPodsInNamespaceToBeReady(namespace)
+		By("Logging in")
+		fly.Login("test", "test", atcEndpoint)
 
-			By("Creating the web proxy")
-			proxySession, atcEndpoint = startPortForwarding(namespace, "service/"+releaseName+"-web", "8080")
+		By("waiting for a running worker")
+		Eventually(func() []Worker {
+			return getRunningWorkers(fly.GetWorkers())
+		}, 2*time.Minute, 10*time.Second).
+			ShouldNot(HaveLen(0))
 
-			By("Logging in")
-			fly.Login("test", "test", atcEndpoint)
+		fly.Run("set-pipeline", "-n",
+			"-c", "../pipelines/task-waiting.yml",
+			"-p", "some-pipeline",
+		)
 
-			Eventually(func() []Worker {
-				return getRunningWorkers(fly.GetWorkers())
-			}, 2*time.Minute, 10*time.Second).
-				ShouldNot(HaveLen(0))
+		fly.Run("unpause-pipeline", "-p", "some-pipeline")
+		fly.Run("trigger-job", "-j", "some-pipeline/simple-job")
 
-			time.Sleep(2 * time.Second)
+		By("waiting container to be created")
+		Eventually(func() bool {
+			containers := fly.GetContainers()
 
-			fly.Run("set-pipeline", "-n", "-c", "../pipelines/"+c.PipelineYamlFile, "-p", "some-pipeline")
-			fly.Run("unpause-pipeline", "-p", "some-pipeline")
-			fly.Run("trigger-job", "-j", "some-pipeline/simple-job")
+			for _, container := range containers {
+				if container.Type == "task" && container.State == "created" {
+					return true
+				}
+			}
 
-			Eventually(func() *gbytes.Buffer {
-				startSession := fly.Start("builds", "-j", "some-pipeline/simple-job")
-				<-startSession.Exited
-				return startSession.Out
-			}, 1*time.Minute, 1*time.Second).Should(gbytes.Say(".*started.*"))
-
-			time.Sleep(2 * time.Second)
-
-			By("deleting the worker pod")
-			Run(nil, "kubectl", "delete", "pod", "--namespace", namespace, releaseName+"-worker-0", "--wait=false")
-
-			By("seeing that the worker state is retiring")
-			Eventually(func() string {
-				workers := fly.GetWorkers()
-				Expect(workers).To(HaveLen(1))
-				return workers[0].State
-			}, 10*time.Second, 2*time.Second).
-				Should(Equal("retiring"))
-
-			By("seeing that there are no workers")
-			Eventually(func() []Worker {
-				return getRunningWorkers(fly.GetWorkers())
-			}, c.WorkerDisappearingTimeout, 1*time.Second).
-				Should(HaveLen(0))
-
-			By("seeing that workers have been revived")
-			Eventually(func() []Worker {
-				return getRunningWorkers(fly.GetWorkers())
-			}, 2*time.Minute, 10*time.Second).
-				ShouldNot(HaveLen(0))
-
-			if c.RerunJob {
-				By("making sure the first build has succeeded")
+			return false
+		}, 2*time.Minute, 10*time.Second).
+			Should(BeTrue())
+	})
+
+	AfterEach(func() {
+		cleanup(releaseName, namespace, proxySession)
+	})
+
+	Context("terminating the worker", func() {
+		JustBeforeEach(func() {
+			Run(nil, "kubectl", "scale", "--namespace", namespace,
+				"statefulset", releaseName+"-worker", "--replicas=0",
+			)
+		})
+
+		Context("gracefully", func() {
+			BeforeEach(func() {
+				gracePeriod = "600"
+			})
+
+			JustBeforeEach(func() {
```

Comment:
Since this is part of the actual test, and there is only one test in this context, shouldn't this `JustBeforeEach` be part of the `It` 🤔 ?